sql/schemachanger: Update mustRetrievePrimaryIndex to find latest pk

spilchen opened the pull request:
Previously, the mustRetrievePrimaryIndex helper function assumed there was always a single primary key when returning an index element. However, during operations such as adding a new column, multiple primary keys can exist temporarily, with only one being the active key. This change updates the helper to correctly identify and return the latest primary key.

This fixes an issue that was affecting ALTER COLUMN TYPE, which relied on this helper.

Epic: CRDB-25314
Closes #134753
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/136373)
<!-- Reviewable:end -->

fqazi reviewed (commented):
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136373)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/alter_column_type` line 1735 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136373#-OCoT-GJ2GQlH-LLnotw:-OCoT-GJ2GQlH-LLnotx:br6q133) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f3626484cc876297f33a4a3f761bd742d5d4bf0/pkg/sql/logictest/testdata/logic_test/alter_column_type#L1735)):*
> ```
> statement ok
> begin;
> ALTER TABLE t_add_col ADD COLUMN c3 TEXT DEFAULT '<default>';
> ```

Does this work with more complex cases like a primary key swap in a txn or dropping and adding columns in the same txn?


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136373)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go` line 1774 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136373#-OCoTbQOA6gUCeC4msFn:-OCoTbQOA6gUCeC4msFo:bu4lmxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f3626484cc876297f33a4a3f761bd742d5d4bf0/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go#L1774)):*
> ```Go
> 
> func mustRetrievePrimaryIndex(b BuildCtx, tableID catid.DescID) *scpb.PrimaryIndex {
> 	return b.QueryByID(tableID).Filter(publicTargetFilter).FilterPrimaryIndex().MustGetOneElement()
> ```

Wonder if we should be using: getLatestPrimaryIndex instead?


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: Update mustRetrievePrimaryIndex to find latest pk

Previously, the mustRetrievePrimaryIndex helper function assumed there
was always a single primary key when returning an index element.
However, during operations such as adding a new column, multiple primary
keys can exist temporarily, with only one being the active key. This
change removes the helper and uses getLatestPrimaryIndex instead, which
correctly identify and return the latest primary key.

This fixes an issue that was affecting ALTER COLUMN TYPE, which relied
on this helper.

Also, as part of this there were a few helpers for elements that weren't
getting the latest versions. Updated the ones to get column default
expression and on update expressions.

Epic: CRDB-25314
Closes #134753
Release note: none

spilchen reviewed (commented):
I found a few other cases that need to filter for the public version. And I opened #136403 for an issue with drop and add column in the same transaction. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136373#-:-OCsCJpS7-CHhoCEKU-W:bj745i5)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go` line 1774 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136373#-OCoTbQOA6gUCeC4msFn:-OCsBnWz6LHtfttUxspy:b-3gl47v) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f3626484cc876297f33a4a3f761bd742d5d4bf0/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go#L1774)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Wonder if we should be using: getLatestPrimaryIndex instead?
</blockquote></details>

Good idea. I got rid of this helper entirely and used `getLatestPrimaryIndex` instead.
___
*[`pkg/sql/logictest/testdata/logic_test/alter_column_type` line 1735 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136373#-OCoT-GJ2GQlH-LLnotw:-OCsBuQc28zdIfKhrpDB:b4jpt0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f3626484cc876297f33a4a3f761bd742d5d4bf0/pkg/sql/logictest/testdata/logic_test/alter_column_type#L1735)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Does this work with more complex cases like a primary key swap in a txn or dropping and adding columns in the same txn?
</blockquote></details>

I tried a few more cases. There were a few other spots that weren't grabbing the correct element version. I couldn't test out the PK swap. If you drop and add a column in the same transaction, we have trouble building a plan. I opened #136403 for this.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Nice job with the extra tests as well!

> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136373#-:-OCsF0KM7Aj8x75a4_Xu:b-2ke8jc)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/alter_column_type` line 1735 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136373#-OCoT-GJ2GQlH-LLnotw:-OCsEqPP3fOlni9YEQfH:bxnhtey) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f3626484cc876297f33a4a3f761bd742d5d4bf0/pkg/sql/logictest/testdata/logic_test/alter_column_type#L1735)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I tried a few more cases. There were a few other spots that weren't grabbing the correct element version. I couldn't test out the PK swap. If you drop and add a column in the same transaction, we have trouble building a plan. I opened #136403 for this.
</blockquote></details>

Nice find! I wonder why that scenario is broken.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
